File: components/test-support/src/test/java/org/cloudfoundry/credhub/helpers/JsonTestHelper.java
package org.cloudfoundry.credhub.helpers;

import java.io.IOException;
import java.util.Set;

import javax.validation.ConstraintViolation;
import javax.validation.Validation;
import javax.validation.Validator;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.cloudfoundry.credhub.util.TimeModuleFactory;
import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.hamcrest.Matcher;

import static com.fasterxml.jackson.databind.PropertyNamingStrategy.SNAKE_CASE;

final public class JsonTestHelper {

  private JsonTestHelper() {
    super();
  }

  private static final ObjectMapper OBJECT_MAPPER = createObjectMapper();

  private static final Validator VALIDATOR = Validation.buildDefaultValidatorFactory()
    .getValidator();

  public static ObjectMapper createObjectMapper() {
    return new ObjectMapper()
      .registerModule(TimeModuleFactory.createTimeModule())
      .setPropertyNamingStrategy(SNAKE_CASE);
  }

  public static byte[] serialize(final Object object) {
    try {
      return OBJECT_MAPPER.writeValueAsBytes(object);
    } catch (final JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }

  public static String serializeToString(final Object object) {
    try {
      return OBJECT_MAPPER.writeValueAsString(object);
    } catch (final JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> T deserialize(final byte[] json, final Class<T> klass) {
    try {
      return OBJECT_MAPPER.readValue(json, klass);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> T deserialize(final String json, final Class<T> klass) {
    try {
      return deserializeChecked(json, klass);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> T deserializeChecked(final String json, final Class<T> klass) throws IOException {
    return OBJECT_MAPPER.readValue(json, klass);
  }

  public static <T> Set<ConstraintViolation<T>> validate(final T original) {
    return VALIDATOR.validate(original);
  }

  public static <T> Set<ConstraintViolation<T>> deserializeAndValidate(final String json,
                                                                       final Class<T> klass) {
    try {
      final T object = OBJECT_MAPPER.readValue(json, klass);
      return VALIDATOR.validate(object);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> Set<ConstraintViolation<T>> deserializeAndValidate(final byte[] json,
                                                                       final Class<T> klass) {
    try {
      final T object = OBJECT_MAPPER.readValue(json, klass);
      return VALIDATOR.validate(object);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static Matcher<ConstraintViolation<?>> hasViolationWithMessage(final String expectedMessage) {
    return new BaseMatcher<ConstraintViolation<?>>() {
      @Override
      public boolean matches(final Object item) {
        final ConstraintViolation<?> violation = (ConstraintViolation<?>) item;
        return violation.getMessage().equals(expectedMessage);
      }

      @Override
      public void describeTo(final Description description) {
        description.appendText("getMessage() should equal ").appendValue(expectedMessage);
      }
    };
  }

  public static JsonNode parse(final String jsonString) throws Exception {
    return OBJECT_MAPPER.readTree(jsonString);
  }
}

File: components/test-support/src/test/kotlin/org/cloudfoundry/credhub/helpers/JsonTestHelper.kt
package org.cloudfoundry.credhub.helpers

import jakarta.validation.ConstraintViolation
import jakarta.validation.Validation
import org.cloudfoundry.credhub.util.TimeModuleFactory
import org.hamcrest.BaseMatcher
import org.hamcrest.Description
import org.hamcrest.Matcher
import tools.jackson.core.JacksonException
import tools.jackson.databind.DeserializationFeature
import tools.jackson.databind.JsonNode
import tools.jackson.databind.PropertyNamingStrategies
import tools.jackson.databind.json.JsonMapper

class JsonTestHelper private constructor() {
    companion object {
        private val OBJECT_MAPPER = createObjectMapper()
        private val VALIDATOR = Validation.buildDefaultValidatorFactory().validator

        @JvmStatic
        fun createObjectMapper(): JsonMapper =
            JsonMapper
                .builder()
                .addModule(TimeModuleFactory.createTimeModule())
                .propertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true)
                .build()

        @JvmStatic
        fun serialize(dataObject: Any): ByteArray? {
            try {
                return OBJECT_MAPPER.writeValueAsBytes(dataObject)
            } catch (e: JacksonException) {
                throw RuntimeException(e)
            }
        }

        @JvmStatic
        fun serializeToString(dataObject: Any): String {
            try {
                return OBJECT_MAPPER.writeValueAsString(dataObject)
            } catch (e: JacksonException) {
                throw RuntimeException(e)
            }
        }

        @JvmStatic
        fun <T> deserialize(
            json: ByteArray,
            klass: Class<T>,
        ): T = OBJECT_MAPPER.readValue(json, klass)

        @JvmStatic
        fun <T> deserialize(
            json: String,
            klass: Class<T>,
        ): T = deserializeChecked(json, klass)

        @JvmStatic
        fun <T> deserializeChecked(
            json: String,
            klass: Class<T>,
        ): T = OBJECT_MAPPER.readValue(json, klass)

        @JvmStatic
        fun <T> validate(original: T): Set<ConstraintViolation<T>> = VALIDATOR.validate(original)

        @JvmStatic
        fun <T> deserializeAndValidate(
            json: String,
            klass: Class<T>,
        ): Set<ConstraintViolation<T>> {
            val dataObject = OBJECT_MAPPER.readValue(json, klass)
            return VALIDATOR.validate(dataObject)
        }

        @JvmStatic
        fun <T> deserializeAndValidate(
            json: ByteArray,
            klass: Class<T>,
        ): Set<ConstraintViolation<T>> {
            val dataObject = OBJECT_MAPPER.readValue(json, klass)
            return VALIDATOR.validate(dataObject)
        }

        @JvmStatic
        fun hasViolationWithMessage(expectedMessage: String): Matcher<ConstraintViolation<*>> {
            return object : BaseMatcher<ConstraintViolation<*>>() {
                override fun matches(item: Any): Boolean {
                    val violation = item as ConstraintViolation<*>
                    return violation.message == expectedMessage
                }

                override fun describeTo(description: Description) {
                    description
                        .appendText("getMessage() should equal ")
                        .appendValue(expectedMessage)
                }
            }
        }

        @JvmStatic
        @Throws(Exception::class)
        fun parse(jsonString: String): JsonNode = OBJECT_MAPPER.readTree(jsonString)
    }
}
